### PR TITLE
OTG Test - RT-3.1: Policy based VRF selection

### LINKED ISSUE #725: OTG: RT-3.1: Policy based VRF selection

### PR DESCRIPTION
**Pull request by octpetre**
OTG conversion of _**RT-3.1: Policy based VRF selection**_ found [here](https://github.com/openconfig/featureprofiles/blob/main/feature/experimental/policy/policy_vrf_selection/ate_tests/base_vrf_selection/base_vrf_selection_test.go)
Changes from #736 included
Fixes #725

### CODE REVIEW

File: feature/experimental/policy/policy_vrf_selection/otg_tests/base_vrf_selection/base_vrf_selection_test.go [hunk 1]
```diff
@@ -132,8 +138,8 @@ func configInterfaceDUT(i *oc.Interface, me, peer *attrs.Attributes, subintfinde
 
 	if vlan != 0 {
 		// Add VLANs
-		singletag := s.GetOrCreateVlan().GetOrCreateMatch().GetOrCreateSingleTagged()
-		singletag.VlanId = ygot.Uint16(vlan)
+		singletag := s.GetOrCreateVlan()
+		singletag.VlanId = telemetry.UnionUint16(vlan)
```
Comment (jasdeep-hundal) [inline feature/experimental/policy/policy_vrf_selection/otg_tests/base_vrf_selection/base_vrf_selection_test.go:142]: Remove/use ygot.

File: feature/experimental/policy/policy_vrf_selection/otg_tests/base_vrf_selection/base_vrf_selection_test.go [hunk 2]
```diff
@@ -132,8 +138,8 @@ func configInterfaceDUT(i *oc.Interface, me, peer *attrs.Attributes, subintfinde
 
 	if vlan != 0 {
 		// Add VLANs
-		singletag := s.GetOrCreateVlan().GetOrCreateMatch().GetOrCreateSingleTagged()
-		singletag.VlanId = ygot.Uint16(vlan)
+		singletag := s.GetOrCreateVlan()
```
Comment (jasdeep-hundal) [inline feature/experimental/policy/policy_vrf_selection/otg_tests/base_vrf_selection/base_vrf_selection_test.go:141]: Why was this change needed? (I would not expect DUT configuration to change.)

File: feature/experimental/policy/policy_vrf_selection/otg_tests/base_vrf_selection/base_vrf_selection_test.go [hunk 3]
```diff
@@ -241,79 +247,109 @@ func applyForwardingPolicy(t *testing.T, ate *ondatra.ATEDevice, ingressPort str
...
+	eth := dstDev.Ethernets().Add().SetName(ateDst.Name + ".Eth")
+	eth.SetPortName(p2.ID()).SetMac(ateDst.MAC)
+	eth.Vlans().Add().SetName(dstDev.Name() + "-VLAN").SetId(int32(vlan10))
+	eth.Ipv4Addresses().Add().SetName(dstDev.Name() + ".IPv4").SetAddress(ateDst.IPv4).SetGateway(dutDst.IPv4).SetPrefix(int32(ateDst.IPv4Len))
+	eth.Ipv6Addresses().Add().SetName(dstDev.Name() + ".IPv6").SetAddress(ateDst.IPv6).SetGateway(dutDst.IPv6).SetPrefix(int32(ateDst.IPv6Len))
+
+	dstDev2 := topo.Devices().Add().SetName(ateDst2.Name)
+	eth2 := dstDev2.Ethernets().Add().SetName(ateDst2.Name + ".Eth")
+	eth2.SetPortName(p2.ID()).SetMac(ateDst2.MAC)
+	eth2.Vlans().Add().SetName(dstDev2.Name() + "-VLAN").SetId(int32(vlan20))
```
Comment (jasdeep-hundal) [inline feature/experimental/policy/policy_vrf_selection/otg_tests/base_vrf_selection/base_vrf_selection_test.go:281]: Could move this to a separate line below this block to make the test-specific config more clear.

File: feature/experimental/policy/policy_vrf_selection/otg_tests/base_vrf_selection/base_vrf_selection_test.go [hunk 4]
```diff
@@ -241,79 +247,109 @@ func applyForwardingPolicy(t *testing.T, ate *ondatra.ATEDevice, ingressPort str
...
+	topo.Ports().Add().SetName(p1.ID())
+	srcDev := topo.Devices().Add().SetName(ateSrc.Name)
+	ethSrc := srcDev.Ethernets().Add().SetName(ateSrc.Name + ".Eth")
+	ethSrc.SetPortName(p1.ID()).SetMac(ateSrc.MAC)
+	ethSrc.Ipv4Addresses().Add().SetName(srcDev.Name() + ".IPv4").SetAddress(ateSrc.IPv4).SetGateway(dutSrc.IPv4).SetPrefix(int32(ateSrc.IPv4Len))
+	ethSrc.Ipv6Addresses().Add().SetName(srcDev.Name() + ".IPv6").SetAddress(ateSrc.IPv6).SetGateway(dutSrc.IPv6).SetPrefix(int32(ateSrc.IPv6Len))
+
+	topo.Ports().Add().SetName(p2.ID())
+	dstDev := topo.Devices().Add().SetName(ateDst.Name)
+	eth := dstDev.Ethernets().Add().SetName(ateDst.Name + ".Eth")
```
Comment (jasdeep-hundal) [inline feature/experimental/policy/policy_vrf_selection/otg_tests/base_vrf_selection/base_vrf_selection_test.go:272]: Rename ethDst (and the one below to ethDst2).

File: feature/experimental/policy/policy_vrf_selection/otg_tests/base_vrf_selection/base_vrf_selection_test.go [hunk 5]
```diff
@@ -241,79 +247,109 @@ func applyForwardingPolicy(t *testing.T, ate *ondatra.ATEDevice, ingressPort str
...
+	eth.SetPortName(p2.ID()).SetMac(ateDst.MAC)
+	eth.Vlans().Add().SetName(dstDev.Name() + "-VLAN").SetId(int32(vlan10))
```
Comment (jasdeep-hundal) [inline feature/experimental/policy/policy_vrf_selection/otg_tests/base_vrf_selection/base_vrf_selection_test.go:274]: Could move this to a separate line below this block to make the test-specific config more clear.

File: feature/experimental/policy/policy_vrf_selection/otg_tests/base_vrf_selection/base_vrf_selection_test.go [hunk 6]
```diff
@@ -241,79 +247,109 @@ func applyForwardingPolicy(t *testing.T, ate *ondatra.ATEDevice, ingressPort str
...
 	time.Sleep(sleepOnChange)
-	ate.Topology().New().StartProtocols(t)
+	ate.OTG().StartProtocols(t)
 	time.Sleep(sleepOnChange)
 }
 
-func configureATE(t *testing.T, ate *ondatra.ATEDevice) []*ondatra.Flow {
-	t.Logf("*** Configuring ATE interfaces ...")
-	topo := ate.Topology().New()
+func configureATE(t *testing.T, ate *ondatra.ATEDevice) []gosnappi.Flow {
```
Comment (jasdeep-hundal) [inline feature/experimental/policy/policy_vrf_selection/otg_tests/base_vrf_selection/base_vrf_selection_test.go:256]: t.Helper

File: feature/experimental/policy/policy_vrf_selection/otg_tests/base_vrf_selection/base_vrf_selection_test.go [hunk 7]
```diff
@@ -241,79 +247,109 @@ func applyForwardingPolicy(t *testing.T, ate *ondatra.ATEDevice, ingressPort str
...
+	if IPinIP {
+		flow.Packet().Add().Ipv4()
+	}
+	flow.Size().SetFixed(512)
+	flow.Rate().SetChoice("percentage").SetPercentage(5)
+	flow.Duration().FixedPackets().SetPackets(1000)
+	return flow
+}
+
+func sendTraffic(t *testing.T, ate *ondatra.ATEDevice) {
```
Comment (jasdeep-hundal) [inline feature/experimental/policy/policy_vrf_selection/otg_tests/base_vrf_selection/base_vrf_selection_test.go:335]: t.Helper

**Review by jasdeep-hundal — CHANGES_REQUESTED** [review]

File: feature/experimental/policy/policy_vrf_selection/otg_tests/base_vrf_selection/base_vrf_selection_test.go (hunk 1 above)
Comment (octpetre) [reply feature/experimental/policy/policy_vrf_selection/otg_tests/base_vrf_selection/base_vrf_selection_test.go:142]: Synced with the ate version

File: feature/experimental/policy/policy_vrf_selection/otg_tests/base_vrf_selection/base_vrf_selection_test.go (hunk 7 above)
Comment (octpetre) [reply feature/experimental/policy/policy_vrf_selection/otg_tests/base_vrf_selection/base_vrf_selection_test.go:335]: Done

File: feature/experimental/policy/policy_vrf_selection/otg_tests/base_vrf_selection/base_vrf_selection_test.go (hunk 4 above)
Comment (octpetre) [reply feature/experimental/policy/policy_vrf_selection/otg_tests/base_vrf_selection/base_vrf_selection_test.go:272]: Done

File: feature/experimental/policy/policy_vrf_selection/otg_tests/base_vrf_selection/base_vrf_selection_test.go (hunk 6 above)
Comment (octpetre) [reply feature/experimental/policy/policy_vrf_selection/otg_tests/base_vrf_selection/base_vrf_selection_test.go:256]: Done

File: feature/experimental/policy/policy_vrf_selection/otg_tests/base_vrf_selection/base_vrf_selection_test.go (hunk 2 above)
Comment (octpetre) [reply feature/experimental/policy/policy_vrf_selection/otg_tests/base_vrf_selection/base_vrf_selection_test.go:141]: This was copied from the ATE test

File: feature/experimental/policy/policy_vrf_selection/otg_tests/base_vrf_selection/base_vrf_selection_test.go (hunk 5 above)
Comment (octpetre) [reply feature/experimental/policy/policy_vrf_selection/otg_tests/base_vrf_selection/base_vrf_selection_test.go:274]: Done

File: feature/experimental/policy/policy_vrf_selection/otg_tests/base_vrf_selection/base_vrf_selection_test.go (hunk 3 above)
Comment (octpetre) [reply feature/experimental/policy/policy_vrf_selection/otg_tests/base_vrf_selection/base_vrf_selection_test.go:281]: done

**Comment by octpetre** [comment]
Waiting for #845 to get merged. Converting to draft

**Comment by octpetre** [comment]
Waiting for #1187 to get merged

**Review by jasdeep-hundal — DISMISSED** [review]

**Comment by jasdeep-hundal** [comment]
@octpetre : Can you rebase once more to pick up the deviation that was added to the ATE test today? (It should apply cleanly).

**Review by jasdeep-hundal — APPROVED** [review]
